contributors/design-proposals/raw-block-pv.md

```diff
@@ -0,0 +1,635 @@
+
+The implemenation plan would be to rename the current makeDevices to makeGPUDevices and create a separate function to add the raw block devices to the option array to be passed to the container runtime. This would iterate on the paths passed in for the pod/container.
+
+Since the future of this in Kubernetes for GPUs and other plug-able devices is migrating to a device plugin architecture, there are 
+still differentiating components of storage that are enough to not to enforce alignment to their convention. Two factors when
+considering the usage of device plugins center around discoverability and topology of devices. Since neither of these are requirements
+for using raw block devices, the legacy method of populating the devices and appending it to the device array is sufficient.
+
+# Implementation Plan, Features & Milesones
+
+Phase 1: v1.8
+Feature: Pre-provisioned PVs to precreated devices 
+
+               Milestone 1: API changes
+
+               Milestone 2: Restricted Access 
+              
+               Milestone 3: Changes to the mounter interface as today it is assumed 'file' as the default.
+               
+               Milestone 4: Expose volumeMode to users via kubectl
+               
+               Milestone 5: Adds enable/disable configuration to securityContext in PSP (Pod Security Policy) similar to hostPath
+               
+               Milestone 6: Validate container runtime options with user specifcations as indicated in UC3
+	       
+	       Milestone 7: Container Runtime changes
+
+Phase 2:  v1.9
+Feature: Discovery of block devices 
+
+                Milestone 1: Dynamically provisioned PVs to dynamically allocated devices
+
+                Milestone 2: Privileged container concerns    
+		
+		Milestone 3: Flex volume updates
+
+Other considerations:
+
+               Reference volume driver change(Attach/Detach logic) for pre-provisioned PVs
+
+               Reference volume driver change(Provision/Delete logic) for dynamic provisioning
+
+## Mounter interface proposed design
+# Plugin changes
+## New BlockVolumeMapper interface proposed design
+
+```
+ type BlockVolumeMapper interface {
+       Volume
+       CanBlockMap() error
+       SetUpDevice() error
+       SetUpDeviceAt(dir string) error
+       GetVolumePath() string
+       GetVolumeType() v1.PersistentVolumeType
+ }
+ type BlockVolumeUnmapper interface {
+       Volume
+       TearDownDevice() error
+       TearDownDeviceAt(dir string) error
+ }
+```
+## Changes for volume definition under pod directory
+
+Currently, volume is mounted to the following path of a kubelet node when the volumes is in-use.
+
+```
+/var/lib/kubelet/pods/<pod uid>/volumes/<plugin name>/<volume name>
+```
+
+If the volume is raw block device, the volume can't be mounted because there isn't a filesystem.
+Instead of volume mount, plugin needs to create symbolic link with `<volume name>` name under the
```

Comment:
Is one symlink enough? How does kubelet find that a raw device is not used in any pod?

Historically, kubelet allowed one volume to be used by any pod on the same node (despite the volume access mode) - it was bind-mounted to any pod's `/var/lib/kubeled/pod/<uid>/volumes`. Before detaching the device, kubelet scanned /proc/mounts and performed detach only when the volume was not referenced by any mount.

If we want to keep the same behavior, there should be a way how to determine if a raw device is still used by any pod. Symlinks won't help here. Maybe hardlinks + checking nr. of references?

Alternatively, we need to prohibit using one raw device by several pods on the same node. I.e. forbid raw block volumes with ReadWriteMany access mode.

Reply:
@jsafrane @msau42 @erinboyd 

>> Symlinks won't help here. Maybe hardlinks + checking nr. of references?

I confirmed that we can't create hardlink  when partition of /dev and /var/lib/kubelet are different. Unfortunately this depends on the system environment, we can't use this way.
```
# ln /dev/loop0 ./
ln: failed to create hard link './loop0' => '/dev/loop0': Invalid cross-device link
```
Can we use these two ways?

(1) Manage number of device references by using number of <pod uid> symlinks.
Symlinks are stored under plugins dir like this. If the number of symlink reaches one, then we can proceed to step (2).
```
Global Mount Path:
/var/lib/kubeled/plugins/<kubernetes.io/<plugin name>/<plugin dependent path>/<pod uid>

ex.
FC plugin: /var/lib/kubelet/plugins/kubernetes.io/fc/WWN-lun-0/<pod uid1>
                                                               <pod uid2>
```
(2) Manage the device status if the device is opened or not.
We have this logic currently. We can re-use this logic for raw block device too.

>> I.e. forbid raw block volumes with ReadWriteMany access mode.

Agree. Since the device doesn't have filesystem, we must forbid ReadWriteMany access mode even if all consumer pods are on the same worker node. This easily cause data corruption.
On the other hands, I think ReadOnlyMany access mode is still useful even the volume is block.

Reply:
> we must forbid ReadWriteMany access mode even if all consumer pods are on the same worker node. This easily cause data corruption.

Well, there is e.g. clustered LVM that handles one iSCSI LUN attached to multiple nodes with RWM just fine. I am not sure if anyone wants to use it in Kubernetes though, it's pretty low level thing.

Just an idea: if we can be sure that /var/lib/kubelet/plugins and /var/lib/kubelet/pods are on the same mount point, we could make a symlink in `/var/lib/kubelet/plugins/...iscsi.../<volume name>` to the device (symlinks across mounts are OK) and then have hardlinks to this symlink from `/var/lib/kubelet/pods/<uid>/...`. Cleanup would work the same way as with filesystem, with checking nr. of hardlinks instead of nr. of mounts. Could it work? It's it too complicated? Too confusing? Filesystem cleanup is pretty complicated too.

Reply:
have we considered modeling block access as an additional set of possible access mode values (e.g. `BlockReadWriteOnce`, `BlockReadOnlyMany`)? 

I see the following benefits to that approach:

* version skewed kubelets would see PV/PVCs with block access mode values and correctly fail, rather than try to handle them incorrectly as filesystem PV/PVCs
* it would let you model a PV that could be accessed as a block device OR as a filesystem device, and let the PVC indicate what type of access it wanted (since PVs can already advertise multiple access modes, PVCs can choose one or more they want, and binding matches them up)
* It would let us omit access modes that don't make sense for block devices (e.g. no `BlockReadWriteMany`)

Reply:
@jsafrane 
I suppose that your proposal is similar to the current logic. so it would be work.
But one difficult thing is that we have to validate if the /var/lib/kubelet/plugins and /var/lib/kubelet/pods are on the same mount point.
Should we validate this at the phase of kubelet start up?

Reply:
@mtanino @screeley44 @jsafrane @liggitt I don't feel like we as a community should be putting unnecessary limitations on what block an and can't do. If outside of kube you can access your block device as a multiple writer, then you should be able to. We should allow the developer to manage what their application can and can't do. There are two distinct use cases where you have block backing the application and data loss is not critical. Logging and visualization applications. Both are willing to lose a little data over having to block on writes.

Reply:
@jsafrane IIUC, we don't do the refcounting way anymore.  Now, we depend on GetVolumeName() to return a globally unique name, and kubelet updates the node status based on the volumes in the actual state, and the attach/detach controller will not call detach until that unique volume name has been unmounted by all pods.

I'm not sure how attach/detach of shared volumes is handled in other volume plugins that do it in the mount/unmount calls though.

Reply:
IIRC kubelet uses `VolumesInUse` and `GetVolumeName()` only for attachable plugins. E.g. portworx can provide block and does not implement attacher interface. So you still need the cleanup somehow.

It's very hard to check this, as VolumeManager is too complicated and obfuscated by layers of indirection :-(.

Reply:
@jsafrane @msau42 @liggitt so...are we discussing if we need cleanup? I think we agreed this would need to be part of the mounting logic. Are we discussing multiple writer for block?

Reply:
We're talking about how to safely detach volumes that don't implement the Detach interface in the cause of multiple pods sharing the device.  It doesn't have to be multi-writer, it's an issue for multi-reader too.

Reply:
@jsafrane I took a look at the rbd plugin `TearDown` implementation. It looks like it just calls `GetMountRefs` and only [detach](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/rbd/disk_manager.go#L111) when refcnt is 1.

I think we should leave the behavior the same as filesystem volumes.  For those plugins that properly implement the Attach/Detach interface, then the controller will take care of not detaching until all references have unmounted it.  For plugins that do the attach/detach as part of mount/unmount, then they need to fall back on refcnting method.

Reply:
@jsafrane @msau42 
Here is my understanding of current volume detach sequence for attachable and non-attachable plugins and proposal.
1. case attachable: Plugins that properly implement the Attach/Detach interface
- Check number of references to device using GetMountRefs().
- Check device status whether the device is opened or not.
- If those two check passed, then MarkDeviceAsUnmounted() is set to the volume, then Attach/Detach controller will detach this volume.
2. case non-attachable: plugins that do the attach/detach as part of mount/unmount
- Check number of references to device using GetMountRefs().
- If the check passed, plugin detach the volume as part of unmount

At the attachable plugin, the purpose of checking ref count is mainly unmounting of the global mount path. But non-attachable replies on only for the ref count way when they unmount and detach volume. We need to care of this when we introduce block volume.

As for the case of block volume, we can't use GetMountRefs(), but we can still check device opened status.
So, here is my alternative proposal for block volume case.

1. case attachable: Plugins that properly implement the Attach/Detach interface
- Check device status whether the device is opened or not.
- If the check passed, then MarkDeviceAsUnmounted() is set to the volume, then Attach/Detach controller will detach this volume.

2. case non-attachable: plugins that do the attach/detach as part of mount/unmount
- NEW..... Check device status whether the device is opened or not ....NEW (This should be provided helper method.)
- If the check passed, plugin detach the volume as part of unmount

Reply:
Did we come to a conclusion on this topic?

Reply:
@lpabon 
not yet. I'm waiting response from @msau42 @jsafrane 
Do you have any suggestions?

Reply:
> Check device status whether the device is opened or not.

I don't have a good answer here. It looks racy:

1. Someone starts a pod.
2. A/d controller attaches its volume.
3. Kubelet maps the volume.
4. Kubelet starts the pod.
5. User removes the pod
6. kubelet checks the device (opens the device), it's not used -> marks it for attaching
7. the pod actually opens the device.

At 6., can we be really sure that all pods that used the volume and all their containers are really dead?

Reply:
@jsafrane 

Thank you for the comment.
Agree. There is small window between "kubelet checks of the device" and "device is opened via container".

>> At 6., can we be really sure that all pods that used the volume and all their containers are really dead?

Now I think this might depends on the type of CRI, and we can't depend on only this way because of race condition.

Reply:
@jsafrane

If the problem is worrying about a/d controller might detach volume while the device is opened due to the race condition mentioned in above steps, I think we have some protection to prevent that.
1. When a/d attaches a volume, it assumes that the volume is used (mounted) and mark it in the actual state. a/d controller will not detach the volume if it is in use unless timeout.
2. When node status is updated by kubelet, controller might mark volume is not used (unmounted) so that it can detach the volume. However, we have a check in the function https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/attachdetach/cache/actual_state_of_world.go#L333. The idea is that if volume is never marked once as used/mounted by kubelet, controller will not directly mark the volume as unused even if node status shows so. So kubelet must have updated the volume as used at least once, and then controller will mark it as unused when the status shows that. This can prevent the race condition mentioned above.
3. The drawback here is that in case pod is added and deleted quickly, and device is not actually used during this time, a/d controller will not immediately detach the volume. Instead it will wait 6 minutes timeout and then issue a detach. This is due to the protection we added in a/d controller. But the protection is necessary because we cannot rely on the status update from kubelet. The update is done periodically and might not represent the real world. (There might be something we can improve at kubelet side so to avoid this unnecessary 6 minutes delay. I plan to work on it after the 1.8 release)

Reply:
We may still need a refcnt check, just like with filesystem volumes because the time that a device is opened/closed is different from the time we call setup/teardown.